### DISCUSSION TITLE
How To Correctly Build  libdatachannel For Android

### DISCUSSION THREAD

**Discussion by TryRevilo** [paullouisageneau/libdatachannel#742] [General]
Hi Ageneau. I've been trying to set build your library [libdatachannel](https://github.com/paullouisageneau/libdatachannel) for an **Android** project. I've tried both build options you gave : 

- [POSIX-compliant operating systems](https://github.com/paullouisageneau/libdatachannel/blob/master/BUILDING.md#posix-compliant-operating-systems-including-linux-and-apple-macos), and
- [Build directly with Make](https://github.com/paullouisageneau/libdatachannel/blob/master/BUILDING.md#build-directly-with-make-linux-only). 

I, however, keep getting the following error when I try to build the project in Android Studio C/C++ : 

`ld: error: ../../../../libs/libdatachannel/libdatachannel.a(wstransport.o) is incompatible with aarch64linux.`

Would you please advise on how to get it working? How can I build it so that I get **ARM builds for Android**? Any way you can help will be greatly appreciated.

Thank you in advance. 

- - - -

_**P.S :** Thanks for such a great library._

**Comment by paullouisageneau** [comment]
Hi, it looks like libdatachannel is not built for the required ARM64 architecture. The easiest way to build it for the right target is to use the [Android NDK toolchain file](https://developer.android.com/ndk/guides/cmake), something like:
```
$ cmake -B build -DUSE_GNUTLS=0 -DUSE_NICE=0 -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=[Path to the NDK]/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_PLATFORM=[min SDK version]
```

**Reply by TryRevilo** [reply]
Hi thanks for the reply Ageneau.

I tried that :  

    cmake -B build \
        -DUSE_GNUTLS=0 -DUSE_NICE=0 \
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_TOOLCHAIN_FILE=/home/rev/Downloads/rev_data/rev_progs/rev_android_studio/rev_ndk/android-ndk-r25b-linux/android-ndk-r25b/build/cmake/android.toolchain.cmake \
        -DANDROID_ABI=arm64-v8a -DANDROID_PLATFORM=21

It builds fine without errors. However, when I :  

> 
> ```
> $ cd build  
> $ make -j2  
> ```

it all goes well untill :  

    3 warnings generated.
    [ 70%] Building CXX object CMakeFiles/datachannel.dir/src/impl/transport.cpp.o
    [ 70%] Building CXX object CMakeFiles/datachannel.dir/src/impl/verifiedtlstransport.cpp.o
    [ 70%] Building CXX object CMakeFiles/datachannel.dir/src/impl/websocket.cpp.o
    [ 72%] Building CXX object CMakeFiles/datachannel.dir/src/impl/websocketserver.cpp.o
    [ 72%] Building CXX object CMakeFiles/datachannel.dir/src/impl/wstransport.cpp.o
    [ 74%] Building CXX object CMakeFiles/datachannel.dir/src/impl/wshandshake.cpp.o
    [ 74%] Linking CXX shared library libdatachannel.so
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-d1_srtp.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-methods.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_cert.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_init.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_lib.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_rsa.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_sess.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-ssl_stat.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libssl.a(libssl-lib-statem.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-a_int.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-bio_lib.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-bio_meth.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-bss_mem.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-bn_lib.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-bn_rand.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-ec_key.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-err.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-e_aes.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-evp_enc.o) is incompatible with aarch64linux
    ld: error: /home/rev/Downloads/my_open_ssl/libs/libcrypto.a(libcrypto-lib-evp_lib.o) is incompatible with aarch64linux
    ld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
    CMakeFiles/datachannel.dir/build.make:934: recipe for target 'libdatachannel.so' failed
    make[2]: *** [libdatachannel.so] Error 1
    CMakeFiles/Makefile2:352: recipe for target 'CMakeFiles/datachannel.dir/all' failed
    make[1]: *** [CMakeFiles/datachannel.dir/all] Error 2
    Makefile:135: recipe for target 'all' failed
    make: *** [all] Error 2

I think it is something with my OpenSSL setup.  

I have tried various OpenSSL Build and Install approaches including :  

- sudo apt-get install libssl-dev

- **[From Git Source Repo](https://github.com/openssl/openssl#build-and-install)**  

>     $ git clone git://git.openssl.org/openssl.git
>     $ ./Configure
>     $ make
>     $ make test

My System : 

`$ lsb_release -a`

> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 18.04.6 LTS
> Release:    18.04
> Codename:   bion

Could you please advise on what I might be still doing wrong or missing?  

Thank you in advance.

**Reply by paullouisageneau** [reply]
You need to [cross-compile OpenSSL for Android](https://wiki.openssl.org/index.php/Android) and for the right ARM architecture, you can't use a build targeting your computer.
